src/capture/image_scaler.cpp

```diff
@@ -0,0 +1,338 @@
+
+	output.one_per_horiz_scale = 1.0f / output.horiz_scale;
+
+	output.width  = static_cast<uint16_t>(input.width * output.horiz_scale);
+	output.height = static_cast<uint16_t>(input.height * output.vert_scale);
+
+	if (input.is_paletted()) {
+		output.pixel_format = PixelFormat::Indexed8;
+	} else {
+		output.pixel_format = PixelFormat::Rgb888;
+	}
+
+	output.curr_row   = 0;
+	output.row_repeat = 0;
+}
+
+void ImageScaler::UpdateOutputParamsUpscale()
+{
+	constexpr auto target_output_height = 1200;
+
+	// Slight fudge factor to push 350-line modes into 4x vertical scaling
+	constexpr auto fudge_offset = 0.1;
+
+	// Calculate integer vertical scaling factor so the
+	// resulting output image height is roughly around
+	// 1200px. See TODO table for further details.
+	output.vert_scale = static_cast<uint8_t>(
+	        round(static_cast<float>(target_output_height) / input.height +
+	              fudge_offset));
+
+	output.vert_scaling_mode = PerAxisScaling::Integer;
+
+	// Determine horizontal scaling factor & scaling mode
+	const auto horiz_scale_fract = input.pixel_aspect_ratio * output.vert_scale;
+
+	output.horiz_scale = static_cast<float>(horiz_scale_fract.ToDouble());
+
+	output.one_per_horiz_scale = static_cast<float>(
+	        horiz_scale_fract.Inverse().ToDouble());
+
+	if (input.double_width) {
+		output.horiz_scale *= 2.0f;
+	}
+	if (input.double_height) {
+		output.vert_scale *= 2;
+	}
+
+	if (is_integer(output.horiz_scale)) {
+		output.horiz_scale = static_cast<float>(
+		        static_cast<uint16_t>(output.horiz_scale));
+
+		output.horiz_scaling_mode = PerAxisScaling::Integer;
+	} else {
+		output.horiz_scaling_mode = PerAxisScaling::Fractional;
+	}
+
+	// Determine scaled output dimensions, taking the double
+	// width/height input flags into account
+	output.width  = static_cast<uint16_t>(input.width * output.horiz_scale);
+	output.height = static_cast<uint16_t>(input.height * output.vert_scale);
+
+	// Determine pixel format
+	const auto only_integer_scaling = (output.horiz_scaling_mode ==
+	                                           PerAxisScaling::Integer &&
+	                                   output.vert_scaling_mode ==
+	                                           PerAxisScaling::Integer);
+
+	if (only_integer_scaling && input.is_paletted()) {
+		output.pixel_format = PixelFormat::Indexed8;
+	} else {
+		output.pixel_format = PixelFormat::Rgb888;
+	}
+
+	output.curr_row   = 0;
+	output.row_repeat = 0;
+}
+
+void ImageScaler::LogParams()
+{
+	auto pixel_format_to_string = [](const PixelFormat pf) -> std::string {
+		switch (pf) {
+		case PixelFormat::Indexed8: return "Indexed8";
+		case PixelFormat::Rgb888: return "RGB888";
+		}
+	};
+
+	auto scale_mode_to_string = [](const PerAxisScaling sm) -> std::string {
+		switch (sm) {
+		case PerAxisScaling::Integer: return "Integer";
+		case PerAxisScaling::Fractional: return "Fractional";
+		}
+	};
+
+	LOG_MSG("ImageScaler params:\n"
```

Comment:
My compiler (GCC 13.1.0, AMD64) complains here:
```
../src/capture/image_scaler.cpp:159:17: warning: format '%lld' expects argument of type 'long long int', but argument 11 has type 'int64_t' {aka 'long int'} [-Wformat=]
```

Reply:
You have one weird compiler, man 😆 I'll check the CI warnings tomorrow... but zero warnings & errors on clang & Visual Studio 2019.